### PR TITLE
More checks before assuming FIM tokens

### LINKED ISSUE #7592: Bug: Some "code" models invoke undefined behavior at load time after #6745
### What happened?

Steps to reproduce:
- Download https://huggingface.co/TheBloke/Mistral-7B-Code-16K-qlora-GGUF/blob/main/mistral-7b-code-16k-qlora.Q4_K_M.gguf (this is related to general.name containing the string "code")
- Build with `-DCMAKE_BUILD_TYPE=Debug` on Linux to enable libstdc++ assertions (makes the UB more obvious)
- Try to load the model (it crashes):
```console
$ build/bin/main -m mistral-7b-code-16k-qlora.Q4_K_M.gguf -ngl 99 -n 0 -p ''
llm_load_print_meta: BOS token        = 1 '<s>'
llm_load_print_meta: EOS token        = 2 '</s>'
llm_load_print_meta: UNK token        = 0 '<unk>'
llm_load_print_meta: PAD token        = 0 '<unk>'
llm_load_print_meta: LF token         = 13 '<0x0A>'
/usr/include/c++/14.1.1/bits/stl_vector.h:1149: std::vector<_Tp, _Alloc>::const_reference std::vector<_Tp, _Alloc>::operator[](size_type) const [with _Tp = llama_vocab::token_data; _Alloc = std::allocator<llama_vocab::token_data>; const_reference = const llama_vocab::token_data&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
[1]    40137 IOT instruction (core dumped)  build/bin/main -m mistral-7b-code-16k-qlora.Q4_K_M.gguf
```

When printing the PRE token it crashes, because the token ID is 32007 (the default for "code" models) but n_vocab is only 32000. Before #6745, the models loads successfully.

### Name and Version

```console
$ build/bin/main --version
version: 2702 (b97bc3966)
built with gcc (GCC) 14.1.1 20240522 for x86_64-pc-linux-gnu
```

### What operating system are you seeing the problem on?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by CISC**
Added a bare minimum of extra checks before assuming `CodeLlama`/`CodeGemma` FIM tokens.

This fixes `Codestral` and any other non-CodeLlama/Gemma model with the word "code" in `general.name`.

BTW, `Codestral` was initially released with wrong vocab (missing FIM tokens), updated vocab was released today, however `Codestral` does not use the `middle` token, so I'm generating GGUFs without the `middle` token, this might require a separate fix in server.cpp /infill route!

The [GGUFs](https://huggingface.co/CISCai/Codestral-22B-v0.1-SOTA-GGUF) are up in case anyone wants to test (and perhaps implement an `spm_infill` argument to switch to Suffix/Prefix/Middle sequence).

Fixes #7592

### CODE REVIEW

**Review by mofosyne — APPROVED** [review]

File: llama.cpp [hunk 1]
```diff
@@ -4560,7 +4560,7 @@ static void llm_load_vocab(
                 [](unsigned char c){ return std::tolower(c); });
 
             if (gen_name.find("code") != std::string::npos) {
-                if (model.arch == LLM_ARCH_LLAMA) {
+                if (model.arch == LLM_ARCH_LLAMA && gen_name.find("llama") != std::string::npos) {
```
Comment (ggerganov) [inline llama.cpp:4563]: This is not a good idea, we need something better

Comment (CISC) [reply llama.cpp]: Ok, turns out that was a little simpler to do after moving the code-block down a little.

**Comment by mofosyne** [comment]
@CISC what's going on with the CI? Can you double check? If its not related to your code you may want to rebase against last known good CI build in master.

**Comment by CISC** [comment]
@mofosyne I have no idea, but the Server checks seem to randomly fail for everyone, not just here.

**Comment by CISC** [comment]
@mofosyne So, syncing seems to have "fixed" it, just the normal failures now, no clue what changed. :)

**Comment by CISC** [comment]
@ggerganov Do the checks look ok now?

SPM infill just got merged in `llama-cpp-python`, so only this issue blocking `Codestral` Fill-in-Middle.

**Comment by ggerganov** [comment]
Let's merge it, but I think we are doing it wrong when it comes to FIM tokens. I haven't looked in the details deeply, but this code definitely looks like a big hack. We have to try to figure out something more robust

**Comment by CISC** [comment]
The main problem is that there's no way to know if
1. there are FIM tokens in the vocab
2. the FIM tokens are used by the model

IMO it is unsafe to make assumptions here, you really need to know that the model in question has the correct tokens and support using them.

**Pull request merged**